src/sinks/datadog/logs/sink.rs

```diff
@@ -158,96 +126,148 @@ impl From<io::Error> for RequestBuildError {
     }
 }
 
+impl From<serde_json::Error> for RequestBuildError {
+    fn from(error: serde_json::Error) -> RequestBuildError {
+        RequestBuildError::Json { error }
+    }
+}
+
 struct LogRequestBuilder {
     default_api_key: Arc<str>,
-    encoding: JsonEncoding,
+    transformer: Transformer,
     compression: Compression,
 }
 
-impl RequestBuilder<(Option<Arc<str>>, Vec<Event>)> for LogRequestBuilder {
-    type Metadata = (Arc<str>, EventFinalizers);
-    type Events = Vec<Event>;
-    type Encoder = JsonEncoding;
-    type Payload = Bytes;
-    type Request = LogApiRequest;
-    type Error = RequestBuildError;
+struct CountingWrite<'a, T> {
+    inner: T,
+    count: &'a std::cell::Cell<usize>,
+}
+
+impl<'a, T> CountingWrite<'a, T> {
+    fn into_inner(self) -> T {
+        self.inner
+    }
+}
 
-    fn compression(&self) -> Compression {
-        self.compression
+#[allow(clippy::disallowed_methods)]
+impl<'a, T: std::io::Write> std::io::Write for CountingWrite<'a, T> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.count.set(self.count.get() + buf.len());
+        self.inner.write(buf)
     }
 
-    fn encoder(&self) -> &Self::Encoder {
-        &self.encoding
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner.flush()
     }
+}
 
-    fn split_input(
+impl LogRequestBuilder {
+    fn build_request(
         &self,
-        input: (Option<Arc<str>>, Vec<Event>),
-    ) -> (Self::Metadata, RequestMetadataBuilder, Self::Events) {
-        let (api_key, mut events) = input;
-        let finalizers = events.take_finalizers();
-        let api_key = api_key.unwrap_or_else(|| Arc::clone(&self.default_api_key));
-        let builder = RequestMetadataBuilder::from_events(&events);
+        mut events: Vec<Event>,
+        api_key: Arc<str>,
+    ) -> Result<Vec<LogApiRequest>, RequestBuildError> {
+        // TODO: this estimated json size seems seems redundant with the one in
+        // RequestMetadataBuilder::from_events
```

Comment:
Is this a TODO you mean to investigate as part of these changes? Or just flagging in general for later.

Reply:
I think this is more of a general note for the future. It's out of scope for this change to modify `RequestMetadataBuilder` at all I think.